src/iceberg/update/update_partition_spec.cc

```diff
@@ -0,0 +1,488 @@
+  // Cannot rename and delete
+  ICEBERG_BUILDER_CHECK(!renames_.contains(std::string(field->name())),
+                        "Cannot rename and delete partition field: {}", field->name());
+
+  deletes_.insert(field->field_id());
+  return *this;
+}
+
+UpdatePartitionSpec& UpdatePartitionSpec::RenameField(const std::string& name,
+                                                      const std::string& new_name) {
+  // Handle existing void field with the new name
+  auto existing_it = name_to_field_.find(new_name);
+  if (existing_it != name_to_field_.end() && IsVoidTransform(*existing_it->second)) {
+    std::string renamed = std::format("{}_{}", existing_it->second->name(),
+                                      existing_it->second->field_id());
+    RenameField(std::string(existing_it->second->name()), renamed);
+  }
+
+  // Cannot rename newly added fields
+  ICEBERG_BUILDER_CHECK(!added_field_names_.contains(name),
+                        "Cannot rename newly added partition field: {}", name);
+
+  auto field_it = name_to_field_.find(name);
+  ICEBERG_BUILDER_CHECK(field_it != name_to_field_.end(),
+                        "Cannot find partition field to rename: {}", name);
+
+  // Cannot delete and rename
+  ICEBERG_BUILDER_CHECK(!deletes_.contains(field_it->second->field_id()),
+                        "Cannot delete and rename partition field: {}", name);
+
+  renames_[name] = new_name;
+  return *this;
+}
+
+Result<UpdatePartitionSpec::ApplyResult> UpdatePartitionSpec::Apply() {
+  ICEBERG_RETURN_UNEXPECTED(CheckErrors());
+
+  // Reserve capacity to avoid reallocations
+  const size_t existing_fields_count = spec_->fields().size();
+  const size_t adds_count = adds_.size();
+  std::vector<PartitionField> new_fields;
+  new_fields.reserve(existing_fields_count + adds_count);
+
+  // Process existing fields
+  for (const auto& field : spec_->fields()) {
+    if (!deletes_.contains(field.field_id())) {
+      // Field is kept, check for rename
+      auto rename_it = renames_.find(std::string(field.name()));
+      if (rename_it != renames_.end()) {
+        new_fields.emplace_back(field.source_id(), field.field_id(), rename_it->second,
+                                field.transform());
+      } else {
+        new_fields.push_back(field);
+      }
+    } else if (format_version_ < 2) {
+      // In V1, deleted fields are replaced with void transform
+      auto rename_it = renames_.find(std::string(field.name()));
+      std::string field_name =
+          rename_it != renames_.end() ? rename_it->second : std::string(field.name());
+      new_fields.emplace_back(field.source_id(), field.field_id(), field_name,
+                              Transform::Void());
+    }
+    // In V2, deleted fields are simply removed
+  }
+
+  // Add new fields
+  new_fields.insert(new_fields.end(), adds_.begin(), adds_.end());
+
+  // Determine the new spec ID
+  int32_t new_spec_id = spec_ ? spec_->spec_id() + 1 : PartitionSpec::kInitialSpecId;
```

Comment:
The logic here and below are not necessary. The right approach is to implement `TableMetadataBuilder::SetDefaultPartitionSpec` and `TableMetadataBuilder::AddPartitionSpec` to let it handle reusing or creating a new spec_id and then call it from here.

@HeartLinked is refactoring `PendingUpdate::Apply` and `Transaction` so we need to wait that PR before making such changes.